app/tests/reader_studies_tests/test_models.py

```diff
@@ -697,15 +698,17 @@ def test_display_set_title():
 
     # Duplication attempt via edit
     ds1.title = ds0.title
-    with pytest.raises(IntegrityError):
-        ds1.save()
+    with transaction.atomic():
```

Comment:
Why are atomic transactions needed now?

Reply:
They were covered by the `transaction=True` in the django_db decorator first. Without it, the integrity errors cause havoc on the follow-up tests. Now that I think about it, there is the alternative to split the tests forcibly and have the IntegrityError context manager as the final call on the test.

Reply:
I think splitting the tests would make for a really awkward set of tests. So I think the atomic transactions are OK?

Reply:
I think that it is better to split the tests out. Remember the three A's (taken from https://automationpanda.com/2020/07/07/arrange-act-assert-a-pattern-for-writing-good-tests/):

1. **Arrange** inputs and targets. Arrange steps should set up the test case. Does the test require any objects or special settings? Does it need to prep a database? Does it need to log into a web app? Handle all of these operations at the start of the test.
2. **Act** on the target behavior. Act steps should cover the main thing to be tested. This could be calling a function or method, calling a REST API, or interacting with a web page. Keep actions focused on the target behavior.
3. **Assert** expected outcomes. Act steps should elicit some sort of response. Assert steps verify the goodness or badness of that response. Sometimes, assertions are as simple as checking numeric or string values. Other times, they may require checking multiple facets of a system. Assertions will ultimately determine if the test passes or fails.

We tend to have tests that do too much, which makes debugging them annoying - and in this case relies on turning transactions on which is slow. Ideally a test tests only one bit of functionality. If you're worried about duplication you can create a scope limited fixture for the tests which encapsulates the set up.